File: app/src/main/java/org/ametro/providers/IconProvider.java
package org.ametro.providers;

import android.content.Context;
import android.content.res.AssetManager;
import android.graphics.drawable.Drawable;

import java.io.IOException;
import java.util.HashMap;
import java.util.HashSet;

public class IconProvider {

    private final AssetManager assetManager;

    private final HashMap<String,Drawable> icons = new HashMap<>();
    private final HashSet<String> assets = new HashSet<>();
    private final Drawable defaultIcon;
    private final String assetPath;

    public IconProvider(Context context, Drawable defaultIcon, String assetPath){
        this.defaultIcon = defaultIcon;
        this.assetPath = assetPath;
        assetManager = context.getAssets();
        try {
            for (String assetName : assetManager.list(assetPath)) {
                assets.add(assetName.toLowerCase());
            }
        }catch(IOException ex){
            // no icons available
        }
    }

    public Drawable getIcon(String iso) {
        Drawable d = icons.get(iso);
        if (d == null) {
            String assetName = iso.toLowerCase() + ".png";
            if (assets.contains(assetName)) {
                try {
                    d = Drawable.createFromStream(
                            assetManager.open(assetPath + "/" + assetName), null);
                } catch (IOException e) {
                    d = defaultIcon;
                }
            } else {
                d = defaultIcon;
            }
            icons.put(iso, d);
        }
        return d;
    }
}

File: app/src/main/java/io/github/romangolovanov/apps/ametro/providers/IconProvider.kt
package io.github.romangolovanov.apps.ametro.providers

import android.content.Context
import android.graphics.drawable.Drawable
import java.io.IOException

class IconProvider(context: Context, private val defaultIcon: Drawable, private val assetPath: String) {

    private val assetManager = context.assets
    private val icons = mutableMapOf<String, Drawable>()
    private val assets = mutableSetOf<String>()

    init {
        try {
            for (assetName in assetManager.list(assetPath)!!) {
                assets.add(assetName.lowercase())
            }
        } catch (ex: IOException) {
            // no icons available
        }
    }

    fun getIcon(iso: String): Drawable {
        return icons.getOrPut(iso) {
            val assetName = iso.lowercase() + ".png"
            if (assets.contains(assetName)) {
                try {
                    Drawable.createFromStream(assetManager.open("$assetPath/$assetName"), null) ?: defaultIcon
                } catch (e: IOException) {
                    defaultIcon
                }
            } else {
                defaultIcon
            }
        }
    }
}
